src/settings/utils/grouping.js

```diff
@@ -9,6 +9,70 @@ export const getKeyBasedArrayGroup = (array, key) => {
     return acc;
   }, {});
 };
+
+const groupTypeCapabilityByResource = (data) => Object.entries(data).reduce((acc, [type, capabilities]) => {
+  if (!acc[type]) {
+    acc[type] = {};
+  }
+  acc[type] = capabilities.reduce((initial, capability) => {
+    if (!initial[capability.resource]) {
+      initial[capability.resource] = [];
+    }
+    initial[capability.resource].push(capability);
+
+    return initial;
+  }, {});
+
+  return acc;
+}, {});
+
+const groupCapabilitiesObjectByTypeAndResource = (groupedTypeByResource) => {
+  const dataType = [];
+  const proceduralType = [];
+  const settingsType = [];
+
+  Object.entries(groupedTypeByResource).forEach(([type, capabilities]) => {
+    Object.entries(capabilities).forEach(([resource, cap]) => {
+      const actionsObject = cap.reduce((acc, item) => {
+        acc[item.action] = item.id;
+        return acc;
+      }, {});
+      if (type === 'data') {
+        dataType.push({ id: cap[0].id, applicationId: cap[0].applicationId, resource, actions: actionsObject });
+      } else if (type === 'procedural') {
+        proceduralType.push({ id: cap[0].id, applicationId: cap[0].applicationId, resource, actions: actionsObject });
+      } else if (type === 'settings') {
+        settingsType.push({ id: cap[0].id, applicationId: cap[0].applicationId, resource, actions: actionsObject });
+      }
+    });
+  });
+  const result = {};
+
+  if (dataType.length) {
+    result.data = dataType;
```

Comment:
Any harm on setting these with the default value? Would be safer to the calling code and at most might need to check `.length` as is typical with arrays.

Reply:
thank you, Ryan. There is no harm, just checking to empty state in usage place